snuba/datasets/metrics_aggregate_processor.py

```diff
@@ -0,0 +1,166 @@
+        raise NotImplementedError
+
+    def process_message(
+        self, message: Mapping[str, Any], metadata: KafkaMessageMetadata
+    ) -> Optional[ProcessedMessage]:
+        if not self._should_process(message):
+            return None
+
+        timestamp = _ensure_valid_date(datetime.utcfromtimestamp(message["timestamp"]))
+        assert timestamp is not None
+
+        keys = []
+        values = []
+        tags = message["tags"]
+        assert isinstance(tags, Mapping)
+        for key, value in sorted(tags.items()):
+            assert key.isdigit()
+            keys.append(int(key))
+            assert isinstance(value, int)
+            values.append(value)
+
+        processed = [
+            {
+                "org_id": _literal(message["org_id"]),
+                "project_id": _literal(message["project_id"]),
+                "metric_id": _literal(message["metric_id"]),
+                "timestamp": _call(
+                    "toStartOfInterval",
+                    (
+                        _call("toDateTime", (_literal(timestamp.isoformat()),),),
+                        _call("toIntervalSecond", (_literal(granularity),)),
+                    ),
+                ),
+                "tags.key": _array_literal(keys),
+                "tags.value": _array_literal(values),
+                **self._process_values(message),
+                "retention_days": _literal(message["retention_days"]),
+                "partition": metadata.partition,
+                "offset": metadata.offset,
+                "granularity": _literal(granularity),
+            }
+            for granularity in self.GRANULARITIES_SECONDS
+        ]
+        return AggregateInsertBatch(processed, None)
+
+
+class SetsAggregateProcessor(MetricsAggregateProcessor):
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        return message["type"] is not None and message["type"] == METRICS_SET_TYPE
+
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        values = message["value"]
+        for v in values:
+            assert isinstance(v, int), "Illegal value in set. Int expected: {v}"
+
+        return {
+            "value": _call(
+                "arrayReduce", (_literal("uniqState"), _array_literal(values)),
+            )
+        }
+
+
+class CounterAggregateProcessor(MetricsAggregateProcessor):
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        return message["type"] is not None and message["type"] == METRICS_COUNTERS_TYPE
+
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        value = message["value"]
+        assert isinstance(
+            value, (int, float)
+        ), "Illegal value for counter value. Int/Float expected {value}"
+
+        return {
+            "value": _call(
+                "arrayReduce", (_literal("sumState"), _array_literal([value]),),
```

Comment:
What about  summing up all the content of the array and passing `sumState(array(THE_ACTUAL_SUM))` so you would pass a much smaller payload. 
Though, let's build this optimization as a follow up and gate it with a runtime flag so we can see if it improves things or make them worse.

Reply:
I think `value` is a single int or float. So there's nothing to sum unless I'm wrong